distributed/src/main/scala/asobu/distributed/service/Extractors.scala

```diff
@@ -0,0 +1,196 @@
+      bodyExtractor: BodyExtractor[LBody]
+    )(implicit
+      gen: LabelledGeneric.Aux[TMessage, TRepr],
+      r: RestOf2.Aux[TRepr, LRemoteExtra, LBody, LParamExtracted],
+      prepend: Prepend.Aux[LParamExtracted, LRemoteExtra, LExtracted],
+      combineTo: CombineTo[LExtracted, LBody, TRepr],
+      rpeb: RouteParamsExtractorBuilder[LParamExtracted]): Extractors.Aux[TMessage, LExtracted, LParamExtracted, LRemoteExtra] = new Extractors[TMessage] {
+
+      type LToSend = LExtracted
+      type LParam = LParamExtracted
+      type LExtra = LRemoteExtra
+      val remoteExtractorDef = RemoteExtractorDef(rpeb, remoteRequestExtractorDefs)
+
+      def localExtract(dr: DistributedRequest[LToSend]): ExtractResult[TMessage] = {
+        import scala.concurrent.ExecutionContext.Implicits.global
+        bodyExtractor.run(dr.body).map { body ⇒
+          val repr = combineTo(dr.extracted, body)
+          gen.from(repr)
+        }
+      }
+    }
+
+    def apply[LParamExtracted <: HList, LBody <: HList, TRepr <: HList](
+      bodyExtractor: BodyExtractor[LBody]
+    )(implicit
+      gen: LabelledGeneric.Aux[TMessage, TRepr],
+      r: RestOf.Aux[TRepr, LBody, LParamExtracted],
+      combineTo: CombineTo[LParamExtracted, LBody, TRepr],
+      rpeb: RouteParamsExtractorBuilder[LParamExtracted]): Extractors.Aux[TMessage, LParamExtracted, LParamExtracted, HNil] = apply(RequestExtractorDefinition.empty, bodyExtractor)
+
+    def apply[TRepr <: HList]()(
+      implicit
+      gen: LabelledGeneric.Aux[TMessage, TRepr],
+      combineTo: CombineTo[TRepr, HNil, TRepr],
+      rpeb: RouteParamsExtractorBuilder[TRepr]
+    ): Extractors.Aux[TMessage, TRepr, TRepr, HNil] = apply(BodyExtractor.empty)
+  }
+
+  def build[TMessage] = new builder[TMessage]
+
+}
+
+@SerialVersionUID(1L)
+case class RemoteExtractorDef[LExtracted <: HList, LParamExtracted <: HList, LRemoteExtra <: HList](
+    routeParamsExtractorBuilder: RouteParamsExtractorBuilder[LParamExtracted],
+    requestExtractorDefinition: RequestExtractorDefinition[LRemoteExtra]
+)(implicit val prepend: Prepend.Aux[LParamExtracted, LRemoteExtra, LExtracted]) {
+
+  /**
+   * Cannot be lazy val which causes it to be mutable
+   *
+   * @return
+   */
+  def extractor: RemoteExtractor[LExtracted] = {
+    val rpe = routeParamsExtractorBuilder()
+    Extractor.zip(rpe.mapF(r ⇒ fromXor(r.v)), requestExtractorDefinition.apply())
+  }
+
+}
+
+object BodyExtractor {
+  val empty = Extractor.empty[AnyContent]
+  def json[T: Reads]: BodyExtractor[T] = Extractor.fromFunction(JsonBodyExtractor.extractBody[T])
+
+}
+
+@implicitNotFound("Cannot construct RouteParamsExtractor out of ${L}")
+trait RouteParamsExtractorBuilder[L <: HList] extends (() ⇒ RouteParamsExtractor[L]) with Serializable
+
+trait MkRouteParamsExtractorBuilder0 {
+
+  //todo: this extract from either path or query without a way to specify one way or another.
+  object kvToKlesili extends Poly1 {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Object name does not match the regular expression '^[A-Z][A-Za-z0-9]*$'](https://www.codacy.com/app/iHeartRadio/asobu/file/2312372375/issues/source?bid=3037259&fileBranchId=3178877#l135)